File: DataStructures/src/main/java/com/zhangdd/tree/ThreadedBinaryTree.java
package com.zhangdd.tree;

/**
 * @author zhangdd on 2021/11/23
 */
public class ThreadedBinaryTree {

    private Node root;

    //由于树是单向的，所以在遍历时需要标识下当前节点的前驱节点
    private Node pre;

    public ThreadedBinaryTree(Node root) {
        this.root = root;
    }

    public void infixThreaded() {
        infixThreaded(root);
    }

    /**
     * 中序线索化
     *
     * @param node 对二叉树进行线索化其实就是对二叉树的节点进行线索化
     */
    public void infixThreaded(Node node) {
        if (null == node) {
            return;
        }
        //处理左节点
        infixThreaded(node.left);

        //处理当前节点 的前驱节点
        if (null == node.left) {
            node.left = pre;
            node.leftType = 1;
        }
        //处理上一个节点 的后继节点
        if (null != pre && null == pre.right) {
            pre.right = node;
            pre.rightType = 1;
        }
        //这里就是标识前驱节点
        pre = node;


        //处理右节点
        infixThreaded(node.right);

    }

    /**
     * 中序遍历线索化二叉树
     */
    public void infixListThreadedTree() {
        infixListThreadedTree(root);
    }

    /**
     * 中序遍历线索化二叉树
     * 首先中序遍历的基本思想是先左节点，在当前节点，在右节点
     * <p>
     * 1.从当前节点开始，判断当前节点的左节点是否是前驱节点类型，如果是，那么当前节点就是第一个节点，且当前节点的左节点还为null
     * 2.输出当前节点
     * 3.判断当前节点的右节点类型是否是后继节点，如果是则可以一直输出（因为在实现线索化时，）
     */
    private void infixListThreadedTree(Node node) {

        if (null == node) {
            return;
        }
        Node temp = node;

        while (null != temp) {

            while (1 != temp.leftType) {
                temp = temp.left;
            }
            System.out.println(temp);


            while (1 == temp.rightType) {
                temp = temp.right;
                System.out.println(temp);
            }

            temp = temp.right;
        }

    }

    public static class Node {
        private final int no;
        private final String name;

        private Node left;
        private Node right;

        //子节点类型，如果为0代表子节点是左子树，如果为1代表前驱节点
        private int leftType;
        //子节点类型，如果为0代表子节点是右子树，如果为1代表前后继节点
        private int rightType;


        public Node(int no, String name) {
            this.no = no;
            this.name = name;
        }

        public Node getLeft() {
            return left;
        }

        public void setLeft(Node left) {
            this.left = left;
        }

        public Node getRight() {
            return right;
        }

        public void setRight(Node right) {
            this.right = right;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "no=" + no +
                    ", name='" + name + '\'' +
                    '}';
        }
    }
}

File: python/DataStructures/tree/ThreadedBinaryTree.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/4

线索化二叉树
"""


class Node:
    def __init__(self, no, name):
        self.no = no
        self.name = name
        self.left = None
        self.right = None
        # 0表示是左子树，1表示是前驱节点
        self.left_type = 0
        # 0表示是右子树，1表示是后继节点
        self.right_type = 0

    def set_left(self, node):
        self.left = node

    def set_right(self, node):
        self.right = node

    def show(self):
        print('no:{},name:{}'.format(self.no, self.name))


class ThreadedBinaryTree:
    def __init__(self, root):
        self.root = root
        self.pre = None

    def infix_threaded(self):
        self.__do_infix_threaded(self.root)

    def __do_infix_threaded(self, node):
        if node is None:
            return
        # 处理左节点
        self.__do_infix_threaded(node.left)

        # 处理当前节点的前驱节点
        if node.left is None:
            node.left = self.pre
            node.left_type = 1

        # 处理上一个节点的 后继节点
        if self.pre is not None and self.pre.right is None:
            self.pre.right = node
            self.pre.right_type = 1

        self.pre = node

        # 处理右节点
        self.__do_infix_threaded(node.right)

    def infix_order_threaded_tree(self):
        self.__do_infix_order_threaded_tree(self.root)

    def __do_infix_order_threaded_tree(self, node):
        if node is None:
            return
        tmp = node
        while tmp is not None:
            while tmp.left_type != 1:
                tmp = tmp.left
            tmp.show()

            while tmp.right_type == 1:
                tmp = tmp.right
                tmp.show()

            tmp = tmp.right


if __name__ == '__main__':
    root = Node(1, 'mick')
    node1 = Node(3, 'jack')
    node2 = Node(6, 'tom')
    node3 = Node(8, 'smith')
    node4 = Node(10, 'marry')
    node5 = Node(14, 'king')

    root.set_left(node1)
    root.set_right(node2)

    node1.set_left(node3)
    node1.set_right(node4)

    node2.set_left(node5)

    tree = ThreadedBinaryTree(root)
    tree.infix_threaded()

    tree.infix_order_threaded_tree()
